tests/roles/cinder_adoption/tasks/main.yaml

```diff
@@ -4,6 +4,14 @@
     {{ oc_header }}
     oc patch openstackcontrolplane openstack --type=merge --patch '{{ cinder_scheduler_api_patch }}'
 
+- name: Set the default volume type
+  ansible.builtin.shell: |
+    {{ shell_header }}
+    {{ oc_header }}
+    CONTROLLER1_SSH="{{ controller1_ssh }}"
+    DEFAULT_TYPE = $($CONTROLLER1_SSH "python3 -c \"import configparser; c = configparser.ConfigParser(); c.read('/var/lib/config-data/puppet-generated/cinder/etc/cinder/cinder.conf'); print(c['DEFAULT']['default_volume_type'])\"")
+    oc patch openstackcontrolplane openstack --type=merge -p '{"spec":{"cinder":{"template":{"customServiceConfig":"[DEFAULT]\ndefault_volume_type = $DEFAULT_TYPE\n"}}}}'
```

Comment:
Hi Francesco,
Setting a new default type might be a valid case for a greenfield deployment but in a upgrade/adoption scenario, we do want to honor the default set in the previous (17.1) deployment hence we want to use the same default.
it would be good to create a new volume type (admin operation) as a valid default but it will be upto the admin to set it because they might have volumes assigned to the old default type.
I noticed that my patch will override the customServiceConfig field but I couldn't think of a better way to set the default (suggestions are welcome!)
Also all these changes will be handled when, during adoption, we will compare the cinder.conf for old and new deployment and set the desired cinder.conf values for RHOSO as secret.
So IMO the purpose of this patch boils down fixing the deployment in a way that we can run tempest in different jobs having different default volume types in the old deployment.

Reply:
Ok, I see your point.
Sorry for my noob questions, I think it makes sense retrieve the previously configured default backend and do the same in this context.
Does this command have the same effect of patching the configuration via `customServiceConfig`:

```
openstack volume type set --property volume_backend_name=<value_coming_from_tripleo> __DEFAULT__
```
If so, I would split the task in 2 parts (for readability at least):
1. retrieve the default backend type
2. run the command mentioned above via `oc rsh openstackclient ...`

Not against what you did there, and I'm asking questions mostly because of a gap in my knowledge, but I would split the two tasks anyway.
This would be my way of configuring the default type according to tripleo, but I might have started from wrong assumptions.

Reply:
I'm happy to answer any questions you have so there are no "noob" questions, feel free to ask!

> Does this command have the same effect of patching the configuration via customServiceConfig:
> openstack volume type set --property volume_backend_name=<value_coming_from_tripleo> __DEFAULT__

Unfortunately no, what the command above does is set the backend name for the __DEFAULT__ type same as the old "tripleo" type. What this means is that any new volumes created from the __DEFAULT__ type will go into the same backend as they used to go when using the "tripleo" type.
But the catch here is, there might still be volumes (that are created pre-adoption) associated with the old "tripleo" type that needs to be retyped to the __DEFAULT__ type before making the __DEFAULT__ as default_volume_type.

Also one other thing here is that tripleo (the deployment tool) deletes the __DEFAULT__ type hence we are failing to find it in the tempest test, otherwise everything would run smoothly so we will need to again create the __DEFAULT__ type to perform the above operations.

Also just to give you an overview of how the default volume type hierarchy works.
1. __DEFAULT__ -> this type is created by cinder and is the last default type cinder resorts to otherwise fails
2. default_volume_type -> the type set with this config option (let's say "tripleo") is the global default for the deployment and has more priority than the __DEFAULT__ type
3. project specific defaults -> we have a specific API to set project level defaults, which don't come to picture here, but have more priority than the default_volume_type type.

The main idea here is to keep the "default_volume_type" value intact during adoption and later it's upto the admin to create a new type and set it as default or use the old default type.

Reply:
Ok,
sorry, it took a bit for me to get back here given the amount of work on bug verification.
I think I understand the use case you are describing and I think it might be useful to get some configuration from the source cloud config file (tripleo) and set the appropriate parameters via `customServiceConfig`.
Right now, in the current testsuite, we only enable `cinder-volume` for the `Ceph` use case, and we know already that `TripleO` sets [1] as backend.
I would only focus on solving that use case with this patch, and plan for a follow up, with the goal in mind of merging [2].
[1] can be changed of course according to what is found in the configuration file you retrieve from `controller-1`, but I think the `Ceph` use case (that is translated to `uni03gamma`), only uses `tripleo_ceph` as default use case atm, and there are no other backends for this topology, so you might just need to statically patch [1] and add `default_volume_type=tripleo_ceph`.
By doing this we can at least validate the `Ceph` use case in the job, and as long as we start working on other unijobs we can extend and make it more dynamic, and based of what you retrieve from the `puppet-generated` config file.
For example, instead of running an  ansible `shell` command, which is not easy to read, maintain and extend from a pure ansible perspective, in that case, we could:

1. Slurp/Extract, from the source cloud (e.g. from controller-1) `cinder.conf` based on the path `/var/lib/config-data/puppet-generated/cinder/etc/cinder/cinder.conf`  (it would be a simple `cat <file>` and its content registered as an ansible varialbe);

2. Extract the relevant information using the `lookup` ansible plugin/filter, for example, the `default_volume_type` can be retrieved with something like:

```yaml
- name: Extract the CephFSID from ceph.conf
  ansible.builtin.set_fact:
     def_vol_type: "{{ lookup('ansible.builtin.ini', 'default_volume_type', section='DEFAULT', file='/path/to/your/cinder.conf') }}"
```
The gathered fact, can then be used in a template to define the default volume type of the source cloud, similar to what has been done for glance/manila with templates [3][4].

3. Patch the ctlplane with the resulting template filled with the variables gathered and extracted from the source `cinder.conf`: for this part, it might be a good idea to not have the cinder template as a default variable [5], but it might be good to render a template and run an `oc patch ...` against a ctlplane passing the `--patch-file=..`.

This would make the flow more clean and it can be easily extended.
As a final thought, I think the current task has a bug.
1. It patches the ctlplane injecting the default_volume_type regardless of the use case
2. Later in the code `cinder-volume` is deployed patching the ctlplane with the `Ceph` configuration [5], overriding what you just did in the previous step, and it reverts the task of the previous step.

Apart from the above, I see a weird syntax issue in the RDO execution in the way the string is injected in `customServiceConfig` [6]:

```
      cinderVolumes:
        volume1:
          customServiceConfig: |
            [tripleo_ceph]
            backend_host=hostgroup
            volume_backend_name=tripleo_ceph
            volume_driver=cinder.volume.drivers.rbd.RBDDriver
            rbd_ceph_conf=/etc/ceph/ceph.conf
            rbd_user=openstack
            rbd_pool=volumes
            rbd_flatten_volume_from_snapshot=False
            report_discard_supported=True
    ...
          customServiceConfig: "[DEFAULT]\ndefault_volume_type = \"tripleo\" \n"
```

But I think we can simply patch [5] to get a `default_volume_type` as per what I described earlier.
Let me know about your thoughts, hope it helps you! 

[1] https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/tests/roles/cinder_adoption/defaults/main.yaml#L48   
[2] https://review.rdoproject.org/r/c/rdo-jobs/+/55125
[3] https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2#L51
[4] https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/tests/roles/glance_adoption/templates/glance_nfs.yaml.j2#L25
[5] https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/tests/roles/cinder_adoption/defaults/main.yaml
[6] https://logserver.rdoproject.org/28/55128/6/check/periodic-adoption-multinode-to-crc-ceph/68da32e/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/crs/openstackcontrolplanes.core.openstack.org/